### PR TITLE
Kimi exploration

### PR DESCRIPTION
**Pull request by gioelegott**
# TODO

- [x] Setup testing infra for single kernels and full pipeline
- [x] Implement `chunk_gate_cumsum`
   - [x] Fix chunk size=128
- [x] Implement `kkt_kda`
- [x] Implement `wy_kda`
- [x] Implement `chunk_h_kda`
- [x] Implement `chunk_o_kda`
- [x] Investigate unstable kernels
- [x] Test end-to-end pto-isa pipeline
- [x] Compare performance of single kda kernels to gdn kernels
- [x] implement megakernel
- [ ] Compare performance of kda and gdn megakernels

# GDN vs KDA: Kernel-by-Kernel Differences

Both models share the same six-stage pipeline structure. The table below maps each
GDN kernel to its KDA counterpart and calls out exactly what changes.

| Stage | GDN kernel | KDA kernel |
|---|---|---|
| 1 | `chunk_cumsum` | `gate_cumsum` |
| 2 | `scaled_dot_kkt` | `kkt_kda` |
| 3 | `solve_tril` | `inversion_kda` |
| 4 | `wy_fast` | `wy_kda` |
| 5 | `chunk_h` | `chunk_h_kda` |
| 6 | `chunk_o` | `chunk_o_kda` |

---

### 1. `chunk_cumsum` vs `gate_cumsum`

Both compute a within-chunk prefix sum of the gate tensor, resetting at every
chunk and sequence boundary. The only difference is the gate shape:

- **GDN** — `g: [B, T, H]`, one scalar per (token, head). Cumsum output is also `[B, T, H]`.
- **KDA** — `g: [B, T, HV, K]`, one value per (token, head, key-dim). Cumsum output is `[B, T, HV, K]`.

Everything else (boundary handling, varlen support) is identical.

---

### 2. `scaled_dot_kkt` vs `kkt_kda`

Both build a strictly lower-triangular intra-chunk matrix by comparing pairs of
key vectors. The difference is how the gate enters the comparison:

- **GDN** — gate is a scalar, applied *after* the dot product:
  ```
  A[r,c] = (k_r · k_c) * exp(g_cs[r] - g_cs[c]) * beta[r]
  ```

- **KDA** — gate is a K-vector, applied *inside* the dot product:
  ```
  L[r,c] = beta[r] * sum_d k[r,d] * k[c,d] * exp(g_cs[r,d] - g_cs[c,d])
  ```

The result is structurally the same (a `[chunk, chunk]` strictly lower-triangular
matrix), but in KDA each key dimension is gated independently before summing,
so the decay is not factorable out of the inner product.

---

### 3. `solve_tril` vs `inversion_kda`

Both invert `(I + M)` where `M` is the strictly lower-triangular matrix from stage 2.
Because `M` is nilpotent, the inverse is a finite polynomial and is exact in
exact arithmetic.

- **GDN** — uses the CubeCore `tri_inverse` NPU kernel (hardware-accelerated).
- **KDA** — `naive_chunk_kda` uses a Neumann recursion:
  `A = -L; for i: A[i,:i] += A[i,:] @ A[:,:i]; INV = A + I`.
  The CPU reference `ref_inversion_kda` uses `torch.linalg.inv` instead, which
  is equivalent for well-conditioned inputs and simpler to read.

---

### 4. `wy_fast` vs `wy_kda`

Both compute two auxiliary tensors used downstream and the formulas are identical:

```
u = INV @ (beta * v)
w = INV @ (beta * exp(g) * k)
```

Beta enters the same way in both models: once as a row-scale when building the KKT
matrix (stage 2), and once as a column-scale here. The full expression is:

```
u = (I + diag(β)·L)^{-1} · diag(β) · v
```

The only difference is that `g` is a scalar per (token, head) in GDN and a K-vector
per (token, head, key-dim) in KDA, so `exp(g) * k` is a per-dimension scale in KDA.

---

### 5. `chunk_h` vs `chunk_h_kda`

Both are the sequential pass: for each chunk in order, snapshot the state S
entering the chunk, subtract the cross-chunk interference to get corrected values,
then update S. The structure is identical; the state decay formula differs:

- **GDN** — scalar decay, one exp per head:
  ```
  v_new   = u - w @ S
  S_new   = exp(g_last) * S  +  k^T @ (v_new * exp(g_last - g_cs))
  ```

- **KDA** — per-K-slot decay, one exp per (head, key-dim):
  ```
  v_corr  = u - w @ S
  k_rest  = k * exp(g_total - g_cs)          # [C, K], restores to reference point
  S_new   = exp(g_total).unsqueeze(-1) * S  +  k_rest^T @ v_corr
  ```

Both output `s_snapshots` (state entering each chunk) and the corrected values
(`v_new` / `v_corr`).

---

### 6. `chunk_o` vs `chunk_o_kda`

Both compute the final output per chunk from the pre-snapshotted state and the
corrected values. Each chunk is independent (no sequential dependency).

Both models factor Aqk identically as `tril(q_eff @ k_eff.T)` with a causal mask,
where:

```
q_eff = q * exp(g_cs)      # absorb the query-side gate
k_eff = k * exp(-g_cs)     # absorb the key-side gate
```

For GDN, `g_cs` is a scalar per (token, head), so `exp(g_cs)` broadcasts over key
dimensions. For KDA, `g_cs` is a K-vector, so the scale is per key-dimension.
In both cases the output is:

```
inter = q_eff @ S               # [c_len, V]
Aqk   = tril(q_eff @ k_eff.T)  # [c_len, c_len], standard outer product + causal mask
o     = inter + Aqk @ v_corr
```

The one remaining structural difference: GDN clips the gate difference to
`exp(min(Δg, 0))`, capping the attention weight at 1 and preventing amplification
when a query position has a later gate than the key. KDA does not clip. This is
safe because the model always produces negative log-gates, so `Δg = g_cs[r] - g_cs[c]
≤ 0` for r ≥ c within a chunk.

---

## Summary of differences

| Property | GDN | KDA |
|---|---|---|
| Gate shape | `[B, T, H]` scalar | `[B, T, HV, K]` vector |
| Gate in KKT | outside dot product | inside dot product |
| Beta application | row in KKT + column in WY | row in KKT + column in WY (identical) |
| Inversion method | CubeCore `tri_inverse` NPU kernel | `linalg.inv` |
| Q/K normalization | none | L2-normalized |
| State shape | `[D, D]` | `[K, V]` |
| State dtype | fp16 | fp32 |
| State decay | scalar `exp(g_last)` | per-K `exp(g_total[k])` |
| Intra-chunk Aqk | `tril(q_eff @ k_eff.T)`, clips Δg ≤ 0 | `tril(q_eff @ k_eff.T)`, no clipping |
| Chunk size | C = 128 | C = 16–64 |

### CODE REVIEW

File: benchmarks/kernel/bench_kda_vs_gdn.py [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+
+    # GDN: [B, T, HV] fp32 — scalar gate per head per token
+    lib_gdn  = load_chunk_cumsum(HV, D, C)
+    g_gdn    = torch.randn(1, T, HV, device=dev, dtype=torch.float32)
+    gs_gdn   = torch.empty_like(g_gdn)
+
+    def run_gdn():
+        lib_gdn.call_kernel(bd, stream, _vp(g_gdn), _vp(gs_gdn), _vp(cu32), batch, T)
+
+    run_gdn(); torch.npu.synchronize()
```
Comment (asobczyk) [inline benchmarks/kernel/bench_kda_vs_gdn.py]:
```suggestion
    run_gdn() 
    torch.npu.synchronize()
```

File: tests/test_kda_single_kernels.py [hunk 2]
```diff
@@ -0,0 +1,816 @@
...
+  - u = A_final @ v = INV @ (beta*v),  w = A_final @ (exp(g)*k) = INV @ (beta*exp(g)*k)
+  - Aqk[r,c] = q_r·(k_c*exp(g_cs[r]-g_cs[c])) for r>=c  (causal, includes diagonal)
+  - output: (q*exp(g_cs)) @ S + Aqk @ (u - w @ S)
+  - state:  S_new[k,:] = exp(g_total[k]) * S[k,:] + sum_c k_rest[c,k]*v_corr[c,:]
+
+Stage device requirements:
+  - cumsum: runs on NPU (requires --device); calls gate_cumsum_kda PTO kernel.
+  - all others: CPU float32 reference only.
+
+Usage::
```
Comment (asobczyk) [inline tests/test_kda_single_kernels.py]:
```suggestion
Usage:
```

File: tests/test_kda_single_kernels.py [hunk 3]
```diff
@@ -0,0 +1,816 @@
...
+def ref_gate_cumsum(g: torch.Tensor, chunk_size: int, cu_seqlens=None) -> torch.Tensor:
+    """Within-chunk cumulative sum of g [B, T, HV, K].
+
+    Resets at chunk boundaries and sequence boundaries.
+    Matches naive_chunk_kda's per-chunk ``g.cumsum(-2)``.
+    """
+    B, T, HV, Kd = g.shape
+    out = torch.zeros(B, T, HV, Kd, dtype=torch.float32)
+    gf = g.float()
+    for bos, eos in _seq_ranges(T, cu_seqlens):
```
Comment (asobczyk) [inline tests/test_kda_single_kernels.py]: `bos` / `eos` not very intuitive. How about `seq_start` / `seq_end`?

File: tests/test_kda_single_kernels.py [hunk 4]
```diff
@@ -0,0 +1,816 @@
...
+        for j in range(0, eos - bos, chunk_size):
+            s, e = bos + j, min(bos + j + chunk_size, eos)
+            out[:, s:e] = gf[:, s:e].cumsum(dim=1)
+    return out
+
+
+# ---------------------------------------------------------------------------
+# Stage 2 – L matrix (gated K×K product, strictly lower triangular)
```
Comment (asobczyk) [inline tests/test_kda_single_kernels.py]:
```suggestion
# Stage 2 – L matrix (gated K @ K.T product, strictly lower triangular part)
```

File: tests/test_kda_single_kernels.py [hunk 5]
```diff
@@ -0,0 +1,816 @@
...
+    B, T, H, _ = A.shape
+    out = torch.zeros(B, T, H, cs, dtype=torch.float32)
+    Af = A.float()
+    for bos, eos in _seq_ranges(T, cu_seqlens):
+        for j in range(0, eos - bos, cs):
+            s, e = bos + j, min(bos + j + cs, eos)
+            v = e - s
+            for h in range(H):
+                Ac = Af[0, s:e, h, :v]  # [v, v], strictly lower triangular
+                inv = torch.linalg.inv(torch.eye(v) + Ac)
```
Comment (asobczyk) [inline tests/test_kda_single_kernels.py]: Here it feels safer to use `numpy.linalg.inv`, in `double` precision, which calls LAPACK under the hood

File: tests/test_kda_single_kernels.py [hunk 6]
```diff
@@ -0,0 +1,816 @@
...
+    return float("nan") if ss_tot < 1e-30 else 1.0 - ss_res / ss_tot
+
+
+def stats_ok(actual: torch.Tensor, expected: torch.Tensor) -> bool:
+    diff = (actual.float() - expected.float()).abs()
+    bound = ATOL + RTOL * expected.float().abs()
+    if (diff <= bound).all():
+        return True
+    mean_abs = float(expected.float().abs().mean())
+    rmse     = float(torch.sqrt((diff.float() ** 2).mean()))
```
Comment (asobczyk) [inline tests/test_kda_single_kernels.py]: Let's check also the frobenius norm error

File: tests/test_kda_single_kernels.py [hunk 7]
```diff
@@ -0,0 +1,816 @@
...
+
+def _r2(y_ref: torch.Tensor, y_pred: torch.Tensor) -> float:
+    ref  = y_ref.detach().float().numpy().ravel().astype(np.float64)
+    pred = y_pred.detach().float().numpy().ravel().astype(np.float64)
+    ss_res = np.sum((ref - pred) ** 2)
+    ss_tot = np.sum((ref - np.mean(ref)) ** 2)
+    return float("nan") if ss_tot < 1e-30 else 1.0 - ss_res / ss_tot
+
+
+def stats_ok(actual: torch.Tensor, expected: torch.Tensor) -> bool:
```
Comment (asobczyk) [inline tests/test_kda_single_kernels.py]: We might want to add `atol` and `rtol` here as input arguments, defaulting to `ATOL` and `RTOL`

File: tests/test_kda_single_kernels.py [hunk 8]
```diff
@@ -0,0 +1,816 @@
...
+
+
+def _rand_cu(n_seq: int, total: int, rng: random.Random) -> list[int]:
+    if n_seq == 1:
+        return [0, total]
+    bnd = sorted(rng.sample(range(1, total), n_seq - 1))
+    return [0] + bnd + [total]
+
+
+def _align_cu(raw: list[int], cs: int) -> list[int]:
```
Comment (asobczyk) [inline tests/test_kda_single_kernels.py:103]: what is `cs`?

File: tests/test_kda_single_kernels.py [hunk 9]
```diff
@@ -0,0 +1,816 @@
...
+    torch.manual_seed(42)
+    q       = torch.randn(1, T, H, K)
+    k       = torch.randn(1, T, H, K)
+    # L2-normalize q and k to match actual KDA usage (model always normalizes).
+    # Unnormalized keys with K=128 produce L matrices with condition numbers ~1e6,
+    # making float32 linalg.inv inaccurate. Normalized keys give L entries ~0
+    # (random orthogonal vectors), keeping (I+L) well-conditioned.
+    q = q / q.norm(dim=-1, keepdim=True).clamp(min=1e-6)
+    k = k / k.norm(dim=-1, keepdim=True).clamp(min=1e-6)
+    v       = torch.randn(1, T, HV, V_DIM)
```
Comment (asobczyk) [inline tests/test_kda_single_kernels.py]:
```suggestion
    v = torch.randn(1, T, HV, V_DIM)
```

File: tests/test_kda_single_kernels.py [hunk 10]
```diff
@@ -0,0 +1,816 @@
...
+    # Keep cumulative gate magnitudes within fp16 range:
+    # for CHUNK=128 the kernel computes exp(-g_cs); with values in (-0.05, 0)
+    # the worst-case cumsum is bounded by ~6.4 so exp stays within fp16 (~600).
+    g_log   = -torch.rand(1, T, HV, K) * 0.05   # values in (-0.05, 0)
```
Comment (asobczyk) [inline tests/test_kda_single_kernels.py]: suggestion: build `g` simply as `rand` and then take its logarithm?

**Review by asobczyk — COMMENTED** [review]
I think it is in a really good initial shape, nice work @gioelegott . I admit I did not look into the kernel .cpp details, but the high-level implementation looks correct. I left some comments, we can open a meta-MR specifically to strengthen the test cases.

File: tests/test_kda_single_kernels.py (hunk 3 above)
Comment (gioelegott) [reply tests/test_kda_single_kernels.py]: This is quite a standard naming convention (also from the SGLang implementation I think)

**Review by asobczyk — APPROVED** [review]
Great work @gioelegott

**Review by learning-chip — APPROVED** [review]
